ntoskrnl/io/iomgr/deviface.c

```diff
@@ -26,192 +27,498 @@ C_ASSERT(sizeof(L"{01234567-89ab-cdef-0123-456789abcdef}") == GUID_STRING_BYTES
+    NTSTATUS Status;
+
+    DeviceKeyHandle = InstanceKeyHandle = NULL;
+
+    /* Separate it into its constituents */
+    Status = IopSeparateSymbolicLink(SymbolicLinkName,
+                                     NULL,
+                                     &MungedString,
+                                     NULL,
+                                     &ReferenceString,
+                                     NULL,
+                                     NULL);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("Failed to separate symbolic link %wZ, Status 0x%08lx\n", SymbolicLinkName, Status);
+        goto Quit;
+    }
+
+    /* Try to open or create device interface keys */
+    if (Create)
+    {
+        Status = IopCreateRegistryKeyEx(&DeviceKeyHandle,
+                                        ClassHandle,
+                                        &MungedString,
+                                        DesiredAccess | KEY_ENUMERATE_SUB_KEYS,
+                                        REG_OPTION_NON_VOLATILE,
+                                        &DeviceKeyDisposition);
     }
     else
     {
-        RtlInitUnicodeString(&ReferenceString, L"#");
+        Status = IopOpenRegistryKeyEx(&DeviceKeyHandle,
+                                      ClassHandle,
+                                      &MungedString,
+                                      DesiredAccess | KEY_ENUMERATE_SUB_KEYS);
     }
 
-    InitializeObjectAttributes(&ObjectAttributes,
-                               &SubKeyName,
-                               OBJ_CASE_INSENSITIVE | OBJ_KERNEL_HANDLE,
-                               *GuidKeyRealP,
-                               NULL);
-    Status = ZwCreateKey(DeviceKeyRealP,
-                         DesiredAccess | KEY_ENUMERATE_SUB_KEYS,
-                         &ObjectAttributes,
-                         0,
-                         NULL,
-                         REG_OPTION_NON_VOLATILE,
-                         NULL);
     if (!NT_SUCCESS(Status))
     {
-        DPRINT1("Failed to open %wZ%wZ\\%wZ Status %x\n", &BaseKeyU, &GuidString, &SubKeyName, Status);
-        goto cleanup;
+        DPRINT1("Failed to create or open %wZ, Status 0x%08lx\n", &MungedString, Status);
+        goto Quit;
     }
 
-    InitializeObjectAttributes(&ObjectAttributes,
-                               &ReferenceString,
-                               OBJ_CASE_INSENSITIVE | OBJ_KERNEL_HANDLE,
-                               *DeviceKeyRealP,
-                               NULL);
-    Status = ZwCreateKey(InstanceKeyRealP,
-                         DesiredAccess,
-                         &ObjectAttributes,
-                         0,
-                         NULL,
-                         REG_OPTION_NON_VOLATILE,
-                         NULL);
+    /* Replace the 1st '\' by '#' in reference string */
+    ASSERT(!ReferenceString.Buffer[0] || ReferenceString.Buffer[0] == L'\\');
+    ReferenceString.Buffer[0] = L'#';
```

Comment:
Ok wait, so now from the ASSERT, I deduce that the ReferenceString buffer may also contain a NULL character as its first character (first condition). If this is correct, `ReferenceString.Length` is quite probably == 0 (since it's the length without the terminating NUL char). But then, I'm not sure you are allowed to unconditionally replace the first character by `#`, because then: 1. What else follows in the buffer? 2. What about `ReferenceString.Length` ?

Reply:
>1. What else follows in the buffer?

As I noted above, after the first `\`/`#` character, starting from the 2nd one, all following buffer contains the actual reference string in case it exists, so then this is also valid and should be also used for the name of the registry key to be created/opened. It can be easily confirmed by looking and comparing symbolic link values and reference string registry key names on Windows XP/2003. Registry key name should usually have the same name as reference string (everything what comes after GUID string + one character), excepting the 1st character must be `#` instead of `\`.
In case ref string is missing, it should be initialized anyway, to properly create a registry key, because it cannot have an empty `''` name, and it shouldn't be a default value, as it is confirmed to be on Windows.
>2. What about ReferenceString.Length ?

Ah, indeed. I probably should update it too. But I doubt how to do it better: can I simply force like `ReferenceString.Length = 1;`, or should I necessarily to only update it either via `RtlInitUnicodeString`, or via `RtlAppendUnicodeToString`, or...? :thinking: It seems to me that the 1st option is also OK, but I'm not sure exactly.